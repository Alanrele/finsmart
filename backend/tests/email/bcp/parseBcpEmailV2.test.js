const fs = require('fs');
const path = require('path');
const { parseBcpEmailV2 } = require('../../../src/lib/email/bcp/parseBcpEmailV2');

function loadFixture(name) {
  const filePath = path.join(__dirname, 'fixtures', name);
  const raw = fs.readFileSync(filePath, 'utf8');
  return JSON.parse(raw);
}

describe('parseBcpEmailV2', () => {
  test('normalizes PEN card purchase email', () => {
    const fixture = loadFixture('card_purchase_pen.json');
    const result = parseBcpEmailV2({
      subject: fixture.subject,
      text: fixture.text,
    });

    expect(result.success).toBe(true);
    expect(result.transaction.confidence).toBeGreaterThanOrEqual(0.75);
    expect(result.transaction).toMatchSnapshot('card_purchase_pen');
  });

  test('normalizes card purchase variations', () => {
    const htmlFixture = loadFixture('card_purchase_html.json');
    const htmlResult = parseBcpEmailV2({
      subject: htmlFixture.subject,
      html: htmlFixture.html,
    });

    expect(htmlResult.success).toBe(true);
    expect(htmlResult.transaction.amount.currency).toBe('PEN');
    expect(htmlResult.transaction.confidence).toBeGreaterThanOrEqual(0.75);

    const usdFixture = loadFixture('card_purchase_usd.json');
    const usdResult = parseBcpEmailV2({
      subject: usdFixture.subject,
      text: usdFixture.text,
    });

    expect(usdResult.success).toBe(true);
    expect(usdResult.transaction.amount.currency).toBe('USD');
    expect(usdResult.transaction.amount.value).toBe('58.30');
  });

  test('returns confidence 0 for non transactional content', () => {
    const fixture = loadFixture('non_transactional.json');
    const result = parseBcpEmailV2({
      subject: fixture.subject,
      text: fixture.text,
    });

    expect(result.success).toBe(false);
    expect(result.confidence).toBe(0);
    expect(result.notes).toContain('template_not_detected');
  });
});

