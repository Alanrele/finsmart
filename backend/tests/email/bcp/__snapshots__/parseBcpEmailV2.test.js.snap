// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseBcpEmailV2 normalizes PEN card purchase email: card_purchase_pen 1`] = `
{
  "accountRef": undefined,
  "amount": {
    "currency": "PEN",
    "value": "125.90",
  },
  "balanceAfter": undefined,
  "cardLast4": "4321",
  "channel": undefined,
  "confidence": 1,
  "exchangeRate": {
    "used": false,
  },
  "location": "Lima",
  "merchant": "SUPERMERCADO LIMA",
  "notes": undefined,
  "occurredAt": "2024-05-12T19:35:00-05:00",
  "operationId": "OP123456",
  "source": "BCP",
  "template": "card_purchase",
}
`;
